crates/flipperzero/src/furi/thread.rs

```diff
@@ -1,7 +1,194 @@
 
+#[cfg(feature = "alloc")]
+use core::{
+    ffi::{c_void, CStr},
+    fmt, str,
+};
+
+#[cfg(feature = "alloc")]
+use alloc::{
+    boxed::Box,
+    ffi::{CString, NulError},
+    string::String,
+    sync::Arc,
+};
+
+#[cfg(feature = "alloc")]
+use crate::furi::sync::Mutex;
+
 use flipperzero_sys as sys;
 
+#[cfg(feature = "alloc")]
+const MIN_STACK_SIZE: usize = 1024;
+
+/// Thread factory, which can be used in order to configure the properties of a new thread.
+#[cfg(feature = "alloc")]
+pub struct Builder {
+    /// Guaranteed to be UTF-8.
+    name: Option<CString>,
+    stack_size: Option<usize>,
+    heap_trace_enabled: Option<bool>,
+}
+
+#[cfg(feature = "alloc")]
+impl Builder {
+    /// Generates the base configuration for spawning a thread, from which configuration
+    /// methods can be chained.
+    pub fn new() -> Self {
+        Self {
+            name: None,
+            stack_size: None,
+            heap_trace_enabled: None,
+        }
+    }
+
+    /// Names the thread-to-be.
+    ///
+    /// Returns an error if the name contains null bytes (`\0`).
+    pub fn name(mut self, name: String) -> Result<Self, NulError> {
+        CString::new(name).map(|name| {
+            self.name = Some(name);
+            self
+        })
+    }
+
+    /// Sets the size of the stack (in bytes) for the new thread.
+    pub fn stack_size(mut self, size: usize) -> Self {
+        self.stack_size = Some(size);
+        self
+    }
+
+    /// Enables heap tracing.
+    ///
+    /// By default, heap tracing is enabled if the Flipper Zero's "heap track mode" is
+    /// either set to "All", or set to "Tree" and the parent's heap tracing is enabled.
+    pub fn enable_heap_trace(mut self) -> Self {
+        self.heap_trace_enabled = Some(true);
+        self
+    }
+
+    /// Spawns a new thread by taking ownership of the `Builder`, and returns its
+    /// [`JoinHandle`].
+    pub fn spawn<F>(self, f: F) -> JoinHandle
+    where
+        F: FnOnce() -> i32,
+        F: Send + 'static,
+    {
+        let Builder {
+            name,
+            stack_size,
+            heap_trace_enabled,
+        } = self;
+        let thread = Thread::new(name, stack_size, heap_trace_enabled);
+
+        // We need to box twice because trait objects are fat pointers, so we need the
+        // second box to obtain a thin pointer to use as the context.
+        type ThreadBody = Box<dyn FnOnce() -> i32>;
+        let thread_body: Box<ThreadBody> = Box::new(Box::new(f));
+        unsafe extern "C" fn run_thread_body(context: *mut c_void) -> i32 {
+            let thread_body = unsafe { Box::from_raw(context as *mut ThreadBody) };
+            thread_body()
+        }
+        let callback: sys::FuriThreadCallback = Some(run_thread_body);
+        let context = Box::into_raw(thread_body);
+
+        let freeman = FreeMan::new();
+
+        struct StateContext {
+            thread: *mut sys::FuriThread,
+            freeman: Arc<Mutex<FreeMan>>,
+        }
```

Comment:
I was thinking about this and was wondering if we could achieve the desired behaviour with just `Arc`.

In `Thread::spawn`:
```rust
pub fn spawn<F>(self, f: F) -> JoinHandle
where
        F: FnOnce() -> i32,
        F: Send + 'static,
{
    ...

    let context = Arc::new(ThreadContext {
        thread: *mut sys::FuriThread,
        // anything else you want to remember (e.g. the UTF-8 thread name)
    });
    let join_handle = JoinHandle {
        context: Arc::clone(&thread_context);
    };
    unsafe {
        sys::furi_thread_set_state_context(thread.thread, Arc::into_raw(context) as *mut c_void);
    }
    
    ...
    
    join_handle
}
```

In `run_state_callback`:
```rust
unsafe extern "C" fn run_state_callback(state: sys::FuriThreadState, context: *mut c_void) {
    if state == sys::FuriThreadState_FuriThreadStateStopped {
        let context = unsafe { Arc::from_raw(context as *mut ThreadContext) };
        if let Some(context) = context.into_inner() {
            // We were able to successfully extract the thread context from `Arc`, thus there is no other refs
            unsafe { sys::furi_thread_free(context.thread) };
        }
    }
}
```

In `ThreadHandle::join`:
```rust
pub fn join(self) -> i32 {
    unsafe {
        sys::furi_thread_join(self.context.thread);

        sys::furi_thread_get_return_code(self.context.thread)
    }
}
```

In `ThreadHandle::drop`:
```rust
fn drop(&mut self) {
    if let Some(context) = context.into_inner() {
        // We were able to successfully extract the thread context from `Arc`, thus there is no other refs
        unsafe { sys::furi_thread_free(context.thread) };
    }
}
```

By using `Arc` as a reference counter, we can avoid the need for keeping track of whether the thread has been stopped or dropped.

I think we could even potentially `ThreadContext` for `Thread::current`, but you would have to call [`Arc::increment_strong_count`](https://doc.rust-lang.org/alloc/sync/struct.Arc.html#method.increment_strong_count) after `Arc::from_raw` to ensure we didn't accidentally free the context when that `Arc`-from-context-pointer dropped.

Reply:
Oh, I didn't notices `Arc::from_inner`, that does work quite nicely. I also originally thought the logic would be more complex than it ended up being.

Reply:
Oh wait, `Arc::into_inner` is an experimental nightly feature, so that would increase the number of things we are blocked on for #58. And `Arc::strong_count` is insufficient because it doesn't lock, so we would need a `Mutex` around the `Arc`, and then we are getting sufficiently complex that the `FreeMan` solution looks fine.